be able to define services with decorators

vangheem opened the pull request:
In attempting to document defining services, I ended up implementing a decorator instead. Everything involved in api.json(lots of stuff packed in there) and zcml seemed a bit to much to ask someone.

Let me know what you think about this approach.

There no scanning step or anything. It's assumed we're going to load services from activated applications.

vangheem committed:
be able to define services with decorators

davisagli commented:
I like this style of configuration, but not separating configuration from importing the module is a problem. What if I want to use some code from a module that defines a service without actually registering that service (i.e. if I want to extend or replace the service)? What if I want to test some code in the module without registering the service?

datakurre commented:
I recall Venusian based approach not having this issue (referring to my venusianconfiguration once again ;). Because with Venusian just importing does not have any side effects unless you also "scan" the module.

vangheem commented:
It only runs the zca registrations if the application is defined in the config.json file. So you can have the views/code without registering it.

davisagli commented:
@vangheem got it, I read the code more carefully and understand now.

bloodbare commented:
Besides I don't like to have multiple ways to do the same thing, is clear that on services its much easy to have a decorator than defining the service on the configuration file. Can we provide an extension on the decorator that title/description/params can be possible arguments defined on the decorator ?

>   "title": "Create a new Portal",
>                 "description": "Creates a new site on the database",
>                 "params": {
>                     "query": {},
>                     "payload": {
>                         "@type": "string",
>                         "title": "string",
>                         "id": "string"
>                     },
>                     "traversal": []
>                 }

vangheem commented:
Is that just for documentation? I don't notice it used anywhere.

In any case, it should work to pass those arguments in--they are seamlessly passed along as the register_service configuration object.

bloodbare commented:
Great we can extend that later, @sneridagh  was working on a service json definition to html for documentation ( there is an endpoint on the Application level API/@apidefinition that exposes the the API contract).

As I see defining services with the directive will not include them on the app_settings['api_definition'] dict. Can we add it there also ?

vangheem commented:
So you want it to automatically add registered services to `app_settings['api_defintion']`?

vangheem commented:
in case anyone is following, Ramon and I discussed this. Results are:

1) we are good on the syntax
2) also store config data to app_settings['api_definition']
3) convert all plone.server services to use this
4) deprecate zcml variant so there aren't 2 ways

gforcada commented on src/plone.server/plone/server/api/__init__.py:
```diff
@@ -0,0 +1,12 @@
+# these imports are done to force loading services
+from . import addons  # noqa
+from . import app  # noqa
+from . import file  # noqa
```
``file`` is a python keyword, it would be wise to rename it to something else at some point, just to prevent hard to debug errors and the likes

vangheem commented:
@gforcada thanks, updated.

@bloodbare okay, it's ready for final review and merge. I've implemented everything we're talking about; however, there is also some cleanup embedded in here to move more things over to interfaces module so we fix some potential circular import issues. It's good for these changes but it might affect people if they are using the old imports in packages.

vangheem commented:
Alright, I was able to figure out the issues with some tests randomly failing. @bloodbare ready to merge! :)
